syntax = "proto2";

package usb2jtag;

message JTAG_COMMAND {
  required JTAGState jtag_start   = 1;
  required JTAGState jtag_end     = 2;
  required uint32 bitcount        = 3;
  required uint32 period          = 4;
  required uint64 payload         = 5;
  required uint32 data_offset     = 6;
  required uint32 data_length     = 7;
  required uint64 data            = 8;
  required bool   data_is_other   = 9;
  required bool   data_is_address = 10;

  enum JTAGState {
    IDLE = 0;
    RESET = 1;
    SHIFT_DR = 2;
    SHIFT_IR = 3;
  }
}

message HL_COMMAND {

  required CMD_TYPE type         = 1;
  repeated JTAG_COMMAND commands = 2;
  required uint32 size                  = 3;

  enum CMD_TYPE {
    READ  = 0;
    WRITE = 1;
    RESET = 2;
  }
}

message PROTOCOL {
  required string target = 1;
  repeated HL_COMMAND commands = 2;
}
